src/core/tracker.cpp

```diff
@@ -0,0 +1,162 @@
+/***************************************************************************
+ tracker.cpp - Tracker
+  ---------------------
+ begin                : 20.02.2020
+ copyright            : (C) 2020 by David Signer
+ email                : david (at) opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QTimer>
+
+#include "tracker.h"
+
+#include "rubberbandmodel.h"
+#include "qgsproject.h"
+
+Tracker::Tracker( QgsVectorLayer *layer, bool visible )
+  : mLayer( layer ),
+    mVisible( visible )
+{
+}
+
+RubberbandModel *Tracker::model() const
+{
+  return mRubberbandModel;
+}
+
+void Tracker::setModel( RubberbandModel *model )
+{
+  if ( mRubberbandModel == model )
+    return;
+  mRubberbandModel = model;
+}
+
+int Tracker::timeInterval() const
+{
+  return mTimeInterval;
+}
+
+void Tracker::setTimeInterval( const int timeInterval )
+{
+  mTimeInterval = timeInterval;
+}
+
+int Tracker::minimumDistance() const
+{
+  return mMinimumDistance;
+}
+
+void Tracker::setConjunction( const bool conjunction )
+{
+  mConjunction = conjunction;
+}
+
+bool Tracker::conjunction() const
+{
+  return mConjunction;
+}
+
+void Tracker::setMinimumDistance( const int minimumDistance )
+{
+  mMinimumDistance = minimumDistance;
+}
+
+void Tracker::trackPosition()
+{
+  if ( std::isnan( model()->currentCoordinate().x() ) || std::isnan( model()->currentCoordinate().y() ) )
+  {
+    return;
+  }
+
+  if ( QgsWkbTypes::hasM( mLayer->wkbType() ) )
+  {
+    if ( !model()->currentPositionTimestamp().isValid() )
+    {
+      return;
+    }
+    double secsSinceStart = 0;
+    if ( mStartPositionTimestamp.isValid() )
+    {
+      secsSinceStart = mStartPositionTimestamp.secsTo( model()->currentPositionTimestamp() );
+    }
+    else
+    {
+      mStartPositionTimestamp = model()->currentPositionTimestamp();
+    }
+    QgsPoint currentCoordinate = model()->currentCoordinate();
+    currentCoordinate.addMValue( secsSinceStart );
+    model()->setCurrentCoordinate( currentCoordinate );
+  }
+
+  model()->addVertex();
+  mTimeIntervalFulfilled = false;
+  mMinimumDistanceFulfilled = false;
+}
+
+void Tracker::positionReceived()
+{
+
+  QVector<QgsPointXY> points = mRubberbandModel->flatPointSequence( QgsProject::instance()->crs() );
+
+  auto pointIt = points.constEnd() - 1;
```

Comment:
is this working for polygon? you need to replicate first point for a polygon.

Reply:
Sorry, I don't understand. On Polygon and on Line we have the point sequence, no? What is there the difference beside that it's stored as polygon or as line?

Reply:
I mean polygon needs to be closed (the last vertex of the enveloppe/ring) is exactly the same as the first one. 
for instance, I added the `closeLine` option here: https://github.com/opengisch/QField/pull/857/files#diff-3b8cead62e95811575d175a1bd3d7528R58
you might need to add it too fo`flatPointSequence`

Reply:
But here we calculate the distance from the current point to the point before. If we would calculate the distance to the first one (because we add the first one as a last one) this won't work. No?
Do we have to care at the this moment if it's a polygon or not?